src/sema/contracts.rs

```diff
@@ -733,6 +733,49 @@ fn check_inheritance(contract_no: usize, ns: &mut ast::Namespace) {
     ns.diagnostics.extend(diagnostics);
 }
 
+/// Given a contract number, check for duplicated names in all public and external functions.
+/// Updates the functions `abi_name` with the mangled name for overloaded functions.
+///
+/// Note: In sema we do not care about the function name too much.
+/// The mangled name is consumed later by the ABI generation.
+pub fn mangle_function_names(contract_no: usize, ns: &mut ast::Namespace) {
+    let mut all_names = HashMap::new();
+    let mangled: HashSet<usize> = ns.contracts[contract_no]
+        .all_functions
+        .keys()
+        .filter_map(|f| {
+            let function = &ns.functions[*f];
+            match (&function.visibility, &function.ty) {
+                (
+                    pt::Visibility::Public(_) | pt::Visibility::External(_),
+                    pt::FunctionTy::Function,
+                ) => all_names
+                    .insert(function.name.clone(), *f)
+                    .map(|other| [*f, other]),
+                _ => None,
+            }
+        })
+        .flatten()
+        .collect();
+
+    for f in mangled {
+        let f = &mut ns.functions[f];
+        f.mangle_name();
+        if let Some(offender) = all_names.get(&f.abi_name) {
+            let message = format!(
+                "mangling the symbol of overloaded function '{}' with signature '{}' results in a new symbol '{}' but this symbol already exists",
+                &f.name, &f.signature, &f.abi_name
+            );
+            ns.diagnostics.push(ast::Diagnostic::error_with_note(
+                f.loc,
+                message,
+                ns.functions[*offender].loc,
+                "this function declaration conflicts with mangled name".into(),
+            ));
+        }
+    }
```

Comment:
If mangled is an option, then this code can be much simpler.

```
loop over functions
   if the name is mangled
     find an unmangled name which matches
       generate an error
```
note that mangled named could also conflict which each other. 

```solidity
import "foo.sol" as X;

struct S { int foo; }

contract C {
       function foo(S s) public {}
       function foo(X.S s) public {}
}
```
Two different functons which both take struct S, but they are different.

Reply:
I can't reproduce this. What was the exact code that made the compiler fail with that assertion?

Reply:
```
glow @ luna in solang on  function-name-mangling [$!?] is 📦 v0.1.13 via 🐍 v3.10.7 via 🦀 v1.64.0 
❯ cat bar.sol 
import "foo.sol" as X;

struct S { int foo; }

contract C {
       function foo(S s) public {}
       function foo(X.S s) public {}
}

glow @ luna in solang on  function-name-mangling [$?] is 📦 v0.1.13 via 🐍 v3.10.7 via 🦀 v1.64.0 
❯ cat foo.sol
struct S { bool s; }


glow @ luna in solang on  function-name-mangling [$?] is 📦 v0.1.13 via 🐍 v3.10.7 via 🦀 v1.64.0 
❯ cargo run -- compile --target substrate bar.sol
   Compiling solang v0.1.13 (/home/glow/code/solang)
    Finished dev [unoptimized + debuginfo] target(s) in 12.05s
     Running `target/debug/solang compile --target substrate bar.sol`
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `Some(FunctionValue { name: "C::C::function::foo__S", address: 0x555a6439a648, is_const: true, is_null: false, llvm_value: "declare internal i32 @\"C::C::function::foo__S\"({ i256 }*)\n", llvm_type: "i32 ({ i256 }*)" })`,
 right: `None`', src/emit/functions.rs:35:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

I'm on the latest commit of the `function-name-mangling` branch (no additional changes)

Reply:
On my branch I get:

```
[sean@xywoleh issue]$ cargo run compile --target substrate  bar.sol
    Finished dev [unoptimized + debuginfo] target(s) in 0.16s
     Running `/home/sean/git/solang/target/debug/solang compile --target substrate bar.sol`
error: function 'foo' overrides function in same contract
  ┌─ /home/sean/git/solang/issue/bar.sol:7:8
  │
6 │        function foo(S s) public {}
  │        ------------------------ previous definition of 'foo'
7 │        function foo(X.S s) public {}
  │        ^^^^^^^^^^^^^^^^^^^^^^^^^^

error: not all contracts are valid
```